src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -483,30 +487,57 @@ void ObjectSynchronizer::enter(Handle obj, BasicLock* lock, JavaThread* current)
   current->inc_held_monitor_count();
 
   if (!useHeavyMonitors()) {
-    markWord mark = obj->mark();
-    if (mark.is_neutral()) {
-      // Anticipate successful CAS -- the ST of the displaced mark must
-      // be visible <= the ST performed by the CAS.
-      lock->set_displaced_header(mark);
-      if (mark == obj()->cas_set_mark(markWord::from_pointer(lock), mark)) {
+    if (UseFastLocking) {
+      // fast-locking does not use the 'lock' parameter.
+      LockStack& lock_stack = current->lock_stack();
+      if (lock_stack.can_push()) {
+        markWord header = obj()->mark_acquire();
+        while (true) {
+          if (header.is_neutral()) {
+            assert(!lock_stack.contains(obj()), "thread must not already hold the lock");
+            // Try to swing into 'fast-locked' state without inflating.
+            markWord locked_header = header.set_fast_locked();
+            markWord witness = obj()->cas_set_mark(locked_header, header);
+            if (witness == header) {
+              // Successfully fast-locked, push object to lock-stack and return.
+              lock_stack.push(obj());
+              return;
+            }
+            // Otherwise retry.
+            header = witness;
+          } else {
+            // Fall-through to inflate-enter.
+            break;
+          }
+        }
+      }
+      // No room on the lock_stack so fall-through to inflate-enter.
+    } else {
+      markWord mark = obj->mark();
```

Comment:
why is it `mark` here but `header` above?

Reply:
Oh I don't know. We are very inconsistent in our nomenclature here and use mark in some places and header in some others (e.g. OM::set_header() or the displaced_header() methods). The name mark is not really fitting and only still exists for historical reasons I believe (when one of the primary functions of the object header is to indicate GC marking?) However, the central data type is still called markWord so I changed my new code paths to use *mark as well. This probably warrants a round of codebase cleanup and consolidation later.